File: clients/ios/Sources/SpectreUI/Components/Content.swift
import SwiftUI
import SpectreCore

struct TextView: View {
    let node: RenderNode
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        // 数値がそのまま来ることがある (単独の式は型が保存される) ので string() で文字列化する
        var text = Text(node.string("text"))
        if let weight = fontWeightOf(node.token("weight", default: "regular")) {
            text = text.fontWeight(weight)
        }
        switch node.token("decoration", default: "none") {
        case "underline": text = text.underline()
        case "strikethrough": text = text.strikethrough()
        default: break
        }

        return text
            .font(theme.font(node.token("typography", default: "bodyMd")))
            .foregroundStyle(theme.color(node.token("color", default: "onSurface"), default: .primary))
            .multilineTextAlignment(textAlignmentOf(node.token("align", default: "start")))
            .lineLimit(node.intOrNil("maxLines"))
            .truncationMode(node.token("truncation", default: "tail") == "middle" ? .middle : .tail)
            .modifier(TextSelectionModifier(selectable: node.bool("selectable", default: false)))
            .frame(maxWidth: .infinity, alignment: frameAlignmentOf(node.token("align", default: "start")))
            .spectreNode(node)
    }
}

struct ImageView: View {
    let node: RenderNode
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        let radius = theme.corner(node.token("radius", default: "none"))
        let placeholderColor = theme.color("surfaceVariant", default: Color.spectreFill)

        // v0.1 は SwiftUI 標準の AsyncImage を使う。ディスクキャッシュと
        // 優先度制御が要るようになったら Nuke へ差し替える (ADR-0001 の想定)。
        AsyncImage(url: URL(string: node.string("url"))) { phase in
            switch phase {
            case .success(let image):
                image
                    .resizable()
                    .aspectRatio(contentMode: contentModeOf(node.token("contentMode", default: "fill")))
            case .failure:
                placeholderColor.overlay(
                    Image(systemName: "photo").foregroundStyle(.secondary)
                )
            default:
                placeholderColor
            }
        }
        .clipShape(RoundedRectangle(cornerRadius: radius, style: .continuous))
        // 装飾画像はスクリーンリーダから隠す
        .accessibilityHidden(node.bool("decorative", default: false))
        .spectreNode(node)
    }
}

struct IconView: View {
    let node: RenderNode
    @Environment(\.spectreTheme) private var theme

    /// body に明示的な return が無いと @ViewBuilder が適用され、代入文が
    /// View 式として解釈されてしまう。寸法の決定は body の外に出す。
    private var size: CGFloat {
        switch node.token("size", default: "md") {
        case "sm": return 16
        case "lg": return 32
        default: return 24
        }
    }

    var body: some View {
        Image(systemName: theme.symbol(node.token("name", default: "")))
            .resizable()
            .scaledToFit()
            .frame(width: size, height: size)
            .foregroundStyle(theme.color(node.token("color", default: "onSurface"), default: .primary))
            .spectreNode(node)
    }
}

struct BadgeView: View {
    let node: RenderNode
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        let tone = node.token("tone", default: "neutral")
        let outlined = node.token("variant", default: "filled") == "outlined"

        let container: Color
        let content: Color
        switch tone {
        case "info":
            container = theme.color("info", default: .blue)
            content = theme.color("onInfo", default: .white)
        case "success":
            container = theme.color("success", default: .green)
            content = theme.color("onSuccess", default: .white)
        case "warning":
            container = theme.color("warning", default: .orange)
            content = theme.color("onWarning", default: .black)
        case "error":
            container = theme.color("error", default: .red)
            content = theme.color("onError", default: .white)
        default:
            container = theme.color("surfaceVariant", default: Color.spectreFill)
            content = theme.color("onSurfaceVariant", default: .secondary)
        }

        let shape = RoundedRectangle(cornerRadius: theme.corner("sm"), style: .continuous)

        return Text(node.string("text"))
            .font(theme.font("caption"))
            .foregroundStyle(outlined ? container : content)
            .padding(.horizontal, theme.space("sm"))
            .padding(.vertical, 2)
            .background {
                if outlined { shape.stroke(container, lineWidth: 1) } else { shape.fill(container) }
            }
            .spectreNode(node)
    }
}

struct ProgressIndicatorView: View {
    let node: RenderNode
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        let value = node.doubleOrNil("value")
        let linear = node.token("kind", default: "circular") == "linear"

        // ProgressViewStyle は静的型なので、型消去せず呼び出し側で分岐する。
        Group {
            switch (linear, value) {
            case (true, .some(let v)):
                ProgressView(value: min(max(v, 0), 1)).progressViewStyle(.linear)
            case (true, .none):
                ProgressView().progressViewStyle(.linear)
            case (false, .some(let v)):
                ProgressView(value: min(max(v, 0), 1)).progressViewStyle(.circular)
            case (false, .none):
                ProgressView().progressViewStyle(.circular)
            }
        }
        .tint(theme.color("primary", default: .accentColor))
        .spectreNode(node)
    }
}

/// 未対応コンポーネントの劣化の最終手段 (docs/compatibility.md §3, ADR-0006)。
///
/// `RenderNode.placeholderType` に対して描く汎用プレースホルダ。省略と違って画面上に
/// 痕跡を残すことで、「何かが表示されないまま失われた」ことを利用者・ホストアプリの
/// どちらから見ても分かるようにする。`componentType` prop に元の未知の型名を積んでいる
/// (`Resolver` 参照)。
struct UnsupportedComponentView: View {
    let node: RenderNode
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        let outline = theme.color("outline", default: Color.spectreSeparator)
        let label = node.a11yLabel() ?? "表示できないコンテンツです (\(node.string("componentType", default: "unknown")))"

        ZStack {
            Image(systemName: theme.symbol("warning"))
                .foregroundStyle(outline)
        }
        .frame(minWidth: 32, minHeight: 32)
        .padding(theme.space("sm"))
        .overlay(
            RoundedRectangle(cornerRadius: theme.corner("sm"), style: .continuous)
                .stroke(outline, lineWidth: 1)
        )
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label)
        .spectreNode(node)
    }
}

/// テキスト選択の可否。
///
/// `.enabled` と `.disabled` は別々の型 (`EnabledTextSelectability` /
/// `DisabledTextSelectability`) なので三項演算子では選べない。分岐で切り替える。
private struct TextSelectionModifier: ViewModifier {
    let selectable: Bool

    @ViewBuilder
    func body(content: Content) -> some View {
        if selectable {
            content.textSelection(.enabled)
        } else {
            content.textSelection(.disabled)
        }
    }
}

File: clients/android/spectre-ui/src/main/kotlin/dev/spectre/ui/components/Content.kt
package dev.spectre.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import dev.spectre.core.RenderNode
import dev.spectre.ui.LocalSpectreTheme
import dev.spectre.ui.a11yHidden
import dev.spectre.ui.a11yLabel
import dev.spectre.ui.bool
import dev.spectre.ui.contentScaleOf
import dev.spectre.ui.fontWeightOf
import dev.spectre.ui.floatOrNull
import dev.spectre.ui.intOrNull
import dev.spectre.ui.overflowOf
import dev.spectre.ui.spectreNode
import dev.spectre.ui.string
import dev.spectre.ui.textAlignOf
import dev.spectre.ui.textDecorationOf
import dev.spectre.ui.token

@Composable
fun TextView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val style = theme.textStyle(node.token("typography", "bodyMd"), MaterialTheme.typography.bodyMedium)
    Text(
        // 数値がそのまま来ることがある (単独の式は型が保存される) ので string() で文字列化する
        text = node.string("text"),
        modifier = modifier.spectreNode(node).applyA11y(node),
        style = style,
        color = theme.color(node.token("color", "onSurface"), MaterialTheme.colorScheme.onSurface),
        textAlign = textAlignOf(node.token("align", "start")),
        fontWeight = fontWeightOf(node.token("weight", "regular")),
        textDecoration = textDecorationOf(node.token("decoration", "none")),
        maxLines = node.intOrNull("maxLines") ?: Int.MAX_VALUE,
        overflow = overflowOf(node.token("truncation", "tail")),
    )
}

@Composable
fun ImageView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val decorative = node.bool("decorative", false)
    val radius = theme.corner(node.token("radius", "none"))
    val placeholderColor = theme.color("surfaceVariant", MaterialTheme.colorScheme.surfaceVariant)

    AsyncImage(
        model = node.string("url"),
        // 装飾画像はスクリーンリーダから隠す
        contentDescription = if (decorative) null else node.a11yLabel(),
        modifier = modifier
            .spectreNode(node)
            .clip(RoundedCornerShape(radius))
            .background(placeholderColor),
        contentScale = contentScaleOf(node.token("contentMode", "fill")),
    )
}

@Composable
fun IconView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val size = when (node.token("size", "md")) {
        "sm" -> 16.dp
        "lg" -> 32.dp
        else -> 24.dp
    }
    Icon(
        imageVector = theme.icon(node.token("name", "")),
        contentDescription = node.a11yLabel(),
        modifier = modifier.spectreNode(node).size(size),
        tint = theme.color(node.token("color", "onSurface"), MaterialTheme.colorScheme.onSurface),
    )
}

@Composable
fun BadgeView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val tone = node.token("tone", "neutral")
    val outlined = node.token("variant", "filled") == "outlined"

    val (container, content) = when (tone) {
        "info" -> theme.color("info", MaterialTheme.colorScheme.tertiary) to
            theme.color("onInfo", MaterialTheme.colorScheme.onTertiary)
        "success" -> theme.color("success", MaterialTheme.colorScheme.primary) to
            theme.color("onSuccess", Color.White)
        "warning" -> theme.color("warning", MaterialTheme.colorScheme.secondary) to
            theme.color("onWarning", Color.Black)
        "error" -> theme.color("error", MaterialTheme.colorScheme.error) to
            theme.color("onError", MaterialTheme.colorScheme.onError)
        else -> theme.color("surfaceVariant", MaterialTheme.colorScheme.surfaceVariant) to
            theme.color("onSurfaceVariant", MaterialTheme.colorScheme.onSurfaceVariant)
    }

    val shape = RoundedCornerShape(theme.corner("sm"))
    val decorated = if (outlined) {
        modifier.spectreNode(node).border(1.dp, container, shape)
    } else {
        modifier.spectreNode(node).background(container, shape)
    }

    Box(decorated.padding(horizontal = theme.space("sm"), vertical = 2.dp)) {
        Text(
            text = node.string("text"),
            style = theme.textStyle("caption", MaterialTheme.typography.labelSmall),
            color = if (outlined) container else content,
        )
    }
}

@Composable
fun ProgressIndicatorView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val value = node.floatOrNull("value")
    val size = when (node.token("size", "md")) {
        "sm" -> 16.dp
        "lg" -> 48.dp
        else -> 32.dp
    }
    val color = theme.color("primary", MaterialTheme.colorScheme.primary)
    val base = modifier.spectreNode(node)

    if (node.token("kind", "circular") == "linear") {
        if (value != null) {
            LinearProgressIndicator(progress = { value.coerceIn(0f, 1f) }, modifier = base, color = color)
        } else {
            LinearProgressIndicator(modifier = base, color = color)
        }
    } else {
        if (value != null) {
            CircularProgressIndicator(
                progress = { value.coerceIn(0f, 1f) },
                modifier = base.size(size),
                color = color,
            )
        } else {
            CircularProgressIndicator(modifier = base.size(size), color = color)
        }
    }
}

/**
 * 未対応コンポーネントの劣化の最終手段 (docs/compatibility.md §3, ADR-0006)。
 *
 * [dev.spectre.core.PLACEHOLDER_NODE_TYPE] に対して描く汎用プレースホルダ。省略と違って
 * 画面上に痕跡を残すことで、「何かが表示されないまま失われた」ことを利用者・ホストアプリの
 * どちらから見ても分かるようにする。`componentType` prop に元の未知の型名を積んでいる
 * ([dev.spectre.core.Resolver] 参照)。
 */
@Composable
fun UnsupportedComponentView(node: RenderNode, modifier: Modifier = Modifier) {
    val theme = LocalSpectreTheme.current
    val outline = theme.color("outline", MaterialTheme.colorScheme.outline)
    val label = node.a11yLabel() ?: "表示できないコンテンツです (${node.string("componentType", "unknown")})"

    Box(
        modifier = modifier
            .spectreNode(node)
            .defaultMinSize(minWidth = 32.dp, minHeight = 32.dp)
            .border(1.dp, outline, RoundedCornerShape(theme.corner("sm")))
            .padding(theme.space("sm"))
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector = theme.icon("warning"),
            contentDescription = null,
            tint = outline,
        )
    }
}

/**
 * a11y のラベル/非表示を Compose のセマンティクスに反映する。
 *
 * アイコンのみのボタンや装飾画像でラベルが欠けているケースはエディタ側の
 * リントで公開前に弾く前提だが、SDK 側でも受け取った指定は必ず反映する。
 */
internal fun Modifier.applyA11y(node: RenderNode): Modifier {
    if (node.a11yHidden()) return this.clearAndSetSemantics { }
    val label = node.a11yLabel() ?: return this
    return this.semantics { contentDescription = label }
}
